pkg/controller/servicemesh/controlplane/grafana.go

```diff
@@ -0,0 +1,128 @@
+package controlplane
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/sha1"
+	"encoding/base64"
+
+	"gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+type GrafanaDatasources struct {
+	ApiVersion  int                  `yaml:"apiVersion"`
+	Datasources []*GrafanaDatasource `yaml:"datasources"`
+}
+
+type GrafanaDatasource struct {
+	Access            string           `yaml:"access"`
+	BasicAuth         bool             `yaml:"basicAuth"`
+	BasicAuthPassword string           `yaml:"basicAuthPassword"`
+	BasicAuthUser     string           `yaml:"basicAuthUser"`
+	Editable          bool             `yaml:"editable"`
+	JsonData          *GrafanaJsonData `yaml:"jsonData"`
+	Name              string           `yaml:"name"`
+	OrgId             int              `yaml:"orgId"`
+	Type              string           `yaml:"type"`
+	Url               string           `yaml:"url"`
+	Version           int              `yaml:"version"`
+}
+
+type GrafanaJsonData struct {
+	TlsSkipVerify bool   `yaml:"tlsSkipVerify"`
+	TimeInterval  string `yaml:"timeInterval"`
+}
+
+func (r *ControlPlaneReconciler) patchPrometheusHtpasswd(object *unstructured.Unstructured) error {
+	r.Log.Info("Creating Prometheus HTPasswd entry", object.GetKind(), object.GetName())
+
+	password, err := generatePassword(255)
+	if err != nil {
+		r.Log.Error(err, "failed to generate the Prometheus password")
+		return err
+	}
+	b64Password := base64.StdEncoding.EncodeToString([]byte(password))
+
+	h := sha1.New()
+	h.Write([]byte(password))
+	authValue := "internal:{SHA}" + base64.StdEncoding.EncodeToString(h.Sum(nil))
+	b64AuthValue := base64.StdEncoding.EncodeToString([]byte(authValue))
+
+	// We store the raw password in order to be able to retrieve it below, when patching Grafana ConfigMap
+	err = unstructured.SetNestedField(object.UnstructuredContent(), b64Password, "data", "rawPassword")
+	if err != nil {
+		r.Log.Error(err, "failed to set prometheus raw password")
+		return err
+	}
+
+	err = unstructured.SetNestedField(object.UnstructuredContent(), b64AuthValue, "data", "auth")
+	if err != nil {
+		r.Log.Error(err, "failed to set prometheus htpasswd entry")
+		return err
+	}
+
+	return nil
+}
+
+func (r *ControlPlaneReconciler) patchGrafanaConfig(object *unstructured.Unstructured) error {
+	r.Log.Info("patching Grafana-Prometheus link", object.GetKind(), object.GetName())
+
+	dsYaml, found, err := unstructured.NestedString(object.UnstructuredContent(), "data", "datasources.yaml")
+	if err != nil {
+		r.Log.Error(err, "could not parse istio-grafana ConfigMap")
+		return err
+	} else if !found {
+		r.Log.Error(nil, "failed to patch istio-grafana ConfigMap")
+		return nil
+	}
+
+	d := &GrafanaDatasources{}
```

Comment:
I think you should be operating against the yaml directly.  While it is easier to access the fields using a typed object, anything that's not defined in the type will be lost, which means round-tripping will fail.  This is why we don't use a typed version of values.yaml in the operator itself.

Reply:
@rcernich Done